src/declaration.d

```diff
@@ -2207,68 +2210,70 @@ public:
-                {
-                    if (i == nestedrefs.dim)
-                    {
-                        nestedrefs.push(fdthis);
-                        break;
-                    }
-                    if (nestedrefs[i] == fdthis)
-                        break;
-                }
-                if (fdthis.ident != Id.require && fdthis.ident != Id.ensure)
-                {
-                    /* __require and __ensure will always get called directly,
-                     * so they never make outer functions closure.
-                     */
-                    //printf("\tfdv = %s\n", fdv->toChars());
-                    //printf("\tfdthis = %s\n", fdthis->toChars());
-                    if (loc.filename)
-                    {
-                        int lv = fdthis.getLevel(loc, sc, fdv);
-                        if (lv == -2) // error
-                            return true;
-                    }
-                    // Add this to fdv->closureVars[] if not already there
-                    for (size_t i = 0; 1; i++)
-                    {
-                        if (i == fdv.closureVars.dim)
-                        {
-                            if (!sc.intypeof && !(sc.flags & SCOPEcompile))
-                                fdv.closureVars.push(this);
-                            break;
-                        }
-                        if (fdv.closureVars[i] == this)
-                            break;
-                    }
-                    //printf("fdthis is %s\n", fdthis->toChars());
-                    //printf("var %s in function %s is nested ref\n", toChars(), fdv->toChars());
-                    // __dollar creates problems because it isn't a real variable Bugzilla 3326
-                    if (ident == Id.dollar)
-                    {
-                        .error(loc, "cannnot use $ inside a function literal");
-                        return true;
-                    }
-                    if (ident == Id.withSym) // Bugzilla 1759
-                    {
-                        ExpInitializer ez = _init.isExpInitializer();
-                        assert(ez);
-                        Expression e = ez.exp;
-                        if (e.op == TOKconstruct || e.op == TOKblit)
-                            e = (cast(AssignExp)e).e2;
-                        return lambdaCheckForNestedRef(e, sc);
-                    }
-                }
+                nestedrefs.push(fdthis);
+                break;
             }
+            if (nestedrefs[i] == fdthis)
+                break;
         }
+
+        /* __require and __ensure will always get called directly,
+         * so they never make outer functions closure.
+         */
+        if (fdthis.ident == Id.require || fdthis.ident == Id.ensure)
+            return false;
+
+        //printf("\tfdv = %s\n", fdv.toChars());
+        //printf("\tfdthis = %s\n", fdthis.toChars());
+        if (loc.filename)
```

Comment:
This `loc.filename` check smells hacky. After this refactoring PR is merged, I'll open a cleanup PR for this.